### ISSUE TITLE
LuaSkin/LuaSkin.h

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/hammerspoon#380]
Extensions in hammerspoon regularly import specific headers files and one of the reasons for including the LuaSkin framework is to ensure externally built modules use the same versions as in the Hammerspoon app itself.  This simplifies moving a potential module in and out of core by minimizing changes to the module source code, and provides a path for future support functions to be made available to all modules.

I'm doing this by adding the following to my Makefiles:

~~~
LUA_INCLUDES ?= $(shell hs -c 'hs.processInfo.bundlePath')/Contents/Frameworks/LuaSkin.framework/Headers
~~~

Ok, so far, though it does require hammerspoon to be running when building.

In the extensions in core, the following is a common begining:

~~~
#import <LuaSkin/LuaSkin.h>
#import "../hammerspoon.h"
~~~

The first can be worked around by adding a symlink in `.../Frameworks/LuaSkin.framework/Headers` with `ln -s . LuaSkin`.

The second by putting `hammerspoon.h` (or a link to it) in `.../Frameworks/LuaSkin.framework`, however hammerspoon.h links to a header in Crashlytics...

At any rate, I know the whole LuaSkin thing is a work in progress and while the above hacks alleviate some of the issues, they may not be the best approach.

Just passing along my experience with the changes.

**Comment by cmsj** [comment]
Agreed this is a weak spot at the moment. I'd love to hear from anyone who has ideas about how this can be done better. We can't be the first app to ship a framework that others want to link against.

For the Crashlytics stuff, perhaps we should wrap that in an #ifdef and just #define CLS_NSLOG to be NSLog if -DUSE_CRASHLYTICS isn't specified?

**Comment by cmsj** [comment]
Apparently the most correct thing to do would be to put LuaSkin.framework into /Library/Frameworks/ so compilers and runtimes find it automatically. That would probably require us to shift to using an Installer package rather than distributing as an .app.

**Comment by asmagill** [comment]
I think I'd rather that this were a symlink (if that will work), even if as a developer, I have to add it by hand... I don't see this as something many will care about, and at least for the near term, it's enough a moving target that I don't want to be messing with differing versions -- I want to know that it's linking against the code in my current Hammerspoon build.

I'll try adding the symlink and removing my attempts at auto-building the path and report back...

I also don't like installers, but that's a personal bias because they make it more difficult for me to decide I want the App in a non-standard location.  Not impossible, just extra steps.

**Comment by cmsj** [comment]
I agree that installers are a bunch of painful overhead for users. The symlink route is interesting, although I wonder if the linker prefers a framework from the app bundle, or /Library/Frameworks. That could be a very painful discovery if the symlink is pointing in the wrong place

**Comment by asmagill** [comment]
Ok, doing the following:

~~~
$ cd /Library/Frameworks/
$ sudo ln -s .../Hammerspoon.app/Contents/Frameworks/LuaSkin.framework .
~~~

and the `#import <LuaSkin/LuaSkin.h>` works fine.

That still leaves us with `#import "../hammerspoon.h"` and Craslytics...  I like the idea of using macros to skip it for external module builds -- Craslytics receiving info about external in-progress modules isn't really helpful -- we don't necessarily have the code; and if it does make it into core eventually, the less tweaks necessary, the better.

As small as it is thus far, it might make sense to go ahead and make that Hammerspoon.framework you were talking about.

**Comment by asmagill** [comment]
FWIW, the following links are already in my /LibraryFrameworks folder:

~~~
lrwxr-xr-x  1 root  wheel   71 Dec 22  2014 AEProfiling.framework -> ../../Applications/Motion.app/Contents/Frameworks/AEProfiling.framework
lrwxr-xr-x  1 root  wheel   74 Dec 22  2014 AERegistration.framework -> ../../Applications/Motion.app/Contents/Frameworks/AERegistration.framework
lrwxr-xr-x  1 root  wheel   74 Dec 22  2014 AudioMixEngine.framework -> ../../Applications/Motion.app/Contents/Frameworks/AudioMixEngine.framework
lrwxr-xr-x  1 root  wheel  106 Jul  9 15:35 LuaSkin.framework -> /opt/amagill/src/hammerspoon/hammerspoon-local/build/Hammerspoon.app/Contents/Frameworks/LuaSkin.framework
lrwxr-xr-x  1 root  wheel   60 Jul  1 02:29 NyxAudioAnalysis.framework -> /System/Library/PrivateFrameworks/Nyx
~~~

**Comment by asmagill** [comment]
More or less addressed by #434 -- should probably update development docs and consider Hammerspoon framework, since hammerspoon.h is slowly getting bigger -- and I have some things in mind that will likely be added over the next week, but this issue, as it stands, is good.
